File: src/__tests__/backup.test.ts
import { describe, it, expect } from 'vitest';
import { buildBackup, isEmptyBackup, parseBackup } from '../utils/backup';
import type { BackupData } from '../utils/backup';

const SAMPLE: BackupData = {
  words: { ubiquitous: 3, run: 'mastered', curiosity: 5 },
  notes: { ubiquitous: ['到处都是', '无处不在'] },
  sentences: {
    ubiquitous: [
      { sentence: 'Screens are ubiquitous.', translation: '屏幕**无处不在**。', createdAt: 1 },
    ],
  },
};

describe('Backup round trip', () => {
  it('writes a versioned file and reads it back unchanged', () => {
    const file = buildBackup(SAMPLE, '2026-09-24T00:00:00.000Z');
    expect(file.version).toBe(2);
    expect(file.exportedAt).toBe('2026-09-24T00:00:00.000Z');

    // Through JSON, exactly as a downloaded file would travel.
    const restored = parseBackup(JSON.parse(JSON.stringify(file)));
    expect(restored).toEqual(SAMPLE);
  });

  it('keeps the three data sets separate', () => {
    const restored = parseBackup(buildBackup(SAMPLE, 'now'));
    expect(Object.keys(restored?.words ?? {})).toHaveLength(3);
    expect(restored?.notes.ubiquitous).toEqual(['到处都是', '无处不在']);
    expect(restored?.sentences.ubiquitous?.[0].translation).toBe('屏幕**无处不在**。');
  });
});

describe('Backup compatibility', () => {
  /**
   * The old export wrote the bare map, while localStorage wrapped the same map in
   * {words: {...}}. Both historically existed, so both have to import.
   */
  it('reads a v1 bare map', () => {
    const restored = parseBackup({ ubiquitous: 3, alice: 'known' });
    expect(restored?.words).toEqual({ ubiquitous: 3, alice: 'mastered' });
    expect(restored?.notes).toEqual({});
    expect(restored?.sentences).toEqual({});
  });

  it('reads a v1 wrapped payload, as found in localStorage', () => {
    const restored = parseBackup({ words: { curiosity: 'learning', wonderland: 2 } });
    expect(restored?.words).toEqual({ curiosity: 5, wonderland: 2 });
    expect(restored?.notes).toEqual({});
  });

  it('keeps importing a v2 file that lost its annotations', () => {
    const restored = parseBackup({ version: 2, words: { run: 1 } });
    expect(restored?.words).toEqual({ run: 1 });
    expect(restored?.sentences).toEqual({});
  });

  it('drops unreadable entries and keeps the rest', () => {
    const restored = parseBackup({
      version: 2,
      words: { good: 2, junk: 'nonsense', worse: { level: 3 } },
      notes: { run: ['跑', 42, null], broken: 'not an array' },
      sentences: { run: [{ sentence: 'He runs.' }, { nothing: true }, 'nope'] },
    });

    expect(restored?.words).toEqual({ good: 2 });
    expect(restored?.notes).toEqual({ run: ['跑'] });
    expect(restored?.sentences).toEqual({ run: [{ sentence: 'He runs.' }] });
  });

  it('rejects things that are not backups at all', () => {
    expect(parseBackup(null)).toBeNull();
    expect(parseBackup('a string')).toBeNull();
    expect(parseBackup([1, 2, 3])).toBeNull();
    expect(parseBackup(42)).toBeNull();
  });

  /**
   * Without the version guard this file would be read as the bare map
   * `{version: 2}` — one bogus word named "version", which importing would then
   * use to replace the reader's entire vocabulary.
   */
  it('rejects a versioned file with no word list', () => {
    expect(parseBackup({ version: 2 })).toBeNull();
    expect(parseBackup({ version: 2, words: 'not an object' })).toBeNull();
  });
});

describe('Empty backup detection', () => {
  it('flags a backup with nothing in it', () => {
    // Importing this would silently wipe the reader's vocabulary.
    expect(isEmptyBackup({ words: {}, notes: {}, sentences: {} })).toBe(true);

    const parsed = parseBackup({ words: {} });
    expect(parsed).not.toBeNull();
    expect(isEmptyBackup(parsed as BackupData)).toBe(true);
  });

  it('does not flag a backup that carries any of the three sets', () => {
    expect(isEmptyBackup(SAMPLE)).toBe(false);
    expect(isEmptyBackup({ words: {}, notes: { run: ['跑'] }, sentences: {} })).toBe(false);
    expect(isEmptyBackup({ words: { run: 5 }, notes: {}, sentences: {} })).toBe(false);
  });
});

File: android/domain/src/test/kotlin/com/inputa/reader/domain/backup/BackupTest.kt
package com.inputa.reader.domain.backup

import com.inputa.reader.domain.model.AnnotationsData
import com.inputa.reader.domain.model.SavedSentence
import com.inputa.reader.domain.model.WordStatus
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * 移植自 Web 版的 src/__tests__/backup.test.ts，逐条对应。
 *
 * 端点全部用**真实的 v2 JSON 文本**做夹具，而不是 Kotlin 对象对 Kotlin 对象 ——
 * 因为这条路径的真实输入是用户在 Web 版里导出、再拷到手机上的那个文件。
 * 只测 Kotlin↔Kotlin 的往返会让「字段名对不上」这类问题整个溜过去。
 */
class BackupTest {

    private fun json(text: String): JsonElement = Json.parseToJsonElement(text)

    private val sampleJson = """
        {
          "version": 2,
          "exportedAt": "2026-09-24T00:00:00.000Z",
          "words": { "ubiquitous": 3, "run": "mastered", "curiosity": 5 },
          "notes": { "ubiquitous": ["到处都是", "无处不在"] },
          "sentences": {
            "ubiquitous": [
              { "sentence": "Screens are ubiquitous.", "translation": "屏幕**无处不在**。", "createdAt": 1 }
            ]
          }
        }
    """.trimIndent()

    // ---------- 主路径：Web 版导出的 v2 文件 ----------

    @Test
    fun `imports a v2 export with all three data sets intact`() {
        val data = Backup.parse(json(sampleJson))
        assertNotNull(data)

        assertEquals(3, data!!.words.size)
        assertEquals(WordStatus.L3, data.words["ubiquitous"])
        assertEquals(WordStatus.MASTERED, data.words["run"])
        assertEquals(WordStatus.L5, data.words["curiosity"])

        assertEquals(listOf("到处都是", "无处不在"), data.notes["ubiquitous"])
        assertEquals("屏幕**无处不在**。", data.sentences["ubiquitous"]?.first()?.translation)
        assertEquals(1L, data.sentences["ubiquitous"]?.first()?.createdAt)
    }

    // ---------- 兼容：两种 v1 形状 ----------

    /**
     * 旧版「导出」写的是裸映射，而 localStorage 里是同一个映射外面包了一层 `{words: {...}}`。
     * 两种历史上都真实存在过，所以两种都得能导入。
     */
    @Test
    fun `reads a v1 bare map and migrates its old status values`() {
        val data = Backup.parse(json("""{"ubiquitous":3,"alice":"known"}"""))
        assertNotNull(data)
        assertEquals(mapOf("ubiquitous" to WordStatus.L3, "alice" to WordStatus.MASTERED), data!!.words)
        assertEquals(emptyMap<String, List<String>>(), data.notes)
        assertEquals(emptyMap<String, List<SavedSentence>>(), data.sentences)
    }

    @Test
    fun `reads a v1 wrapped payload as found in localStorage`() {
        val data = Backup.parse(json("""{"words":{"curiosity":"learning","wonderland":2}}"""))
        assertNotNull(data)
        assertEquals(mapOf("curiosity" to WordStatus.L5, "wonderland" to WordStatus.L2), data!!.words)
        assertEquals(emptyMap<String, List<String>>(), data.notes)
    }

    @Test
    fun `keeps importing a v2 file that lost its annotations`() {
        val data = Backup.parse(json("""{"version":2,"words":{"run":1}}"""))
        assertNotNull(data)
        assertEquals(mapOf("run" to WordStatus.L1), data!!.words)
        assertEquals(emptyMap<String, List<SavedSentence>>(), data.sentences)
    }

    @Test
    fun `drops unreadable entries and keeps the rest`() {
        val data = Backup.parse(
            json(
                """
                {
                  "version": 2,
                  "words": { "good": 2, "junk": "nonsense", "worse": { "level": 3 } },
                  "notes": { "run": ["跑", 42, null], "broken": "not an array" },
                  "sentences": { "run": [{ "sentence": "He runs." }, { "nothing": true }, "nope"] }
                }
                """.trimIndent(),
            ),
        )
        assertNotNull(data)
        assertEquals(mapOf("good" to WordStatus.L2), data!!.words)
        assertEquals(mapOf("run" to listOf("跑")), data.notes)
        assertEquals(mapOf("run" to listOf(SavedSentence("He runs."))), data.sentences)
    }

    // ---------- 拒绝「不是本应用的备份」 ----------

    @Test
    fun `rejects things that are not backups at all`() {
        assertNull(Backup.parse(null))
        assertNull(Backup.parse(json("\"a string\"")))
        assertNull(Backup.parse(json("[1,2,3]")))
        assertNull(Backup.parse(json("42")))
    }

    /**
     * 没有这道闸，这个文件会被当成裸映射读成 `{version: 2}` —— 一个名叫 "version"、
     * 熟练度 2 的假词，导入时用它覆盖掉读者的整本词库。
     */
    @Test
    fun `rejects a versioned file with no word list`() {
        assertNull(Backup.parse(json("""{"version":2}""")))
        assertNull(Backup.parse(json("""{"version":2,"words":"not an object"}""")))
    }

    // ---------- 空备份 ----------

    @Test
    fun `flags a backup with nothing in it`() {
        // 导入它就是静默清空读者的词库。
        assertEquals(true, Backup.isEmpty(BackupData(emptyMap(), AnnotationsData())))

        val parsed = Backup.parse(json("""{"words":{}}"""))
        assertNotNull(parsed)
        assertEquals(true, Backup.isEmpty(parsed!!))
    }

    @Test
    fun `does not flag a backup that carries any of the three sets`() {
        val full = Backup.parse(json(sampleJson))!!
        assertEquals(false, Backup.isEmpty(full))

        assertEquals(
            false,
            Backup.isEmpty(BackupData(emptyMap(), AnnotationsData(notes = mapOf("run" to listOf("跑"))))),
        )
        assertEquals(
            false,
            Backup.isEmpty(BackupData(mapOf("run" to WordStatus.L5), AnnotationsData())),
        )
    }

    // ---------- 导出（第二阶段才接 UI，形状现在就钉住） ----------

    @Test
    fun `build stamps the version and the caller supplied timestamp`() {
        val file = Backup.build(BackupData(emptyMap(), AnnotationsData()), "2026-09-24T00:00:00.000Z")
        assertEquals(2, file.version)
        assertEquals("2026-09-24T00:00:00.000Z", file.exportedAt)
        assertEquals(Backup.EXPORT_VERSION, file.version)
    }
}
